modules/raft/src/main/java/org/apache/ignite/raft/jraft/core/FSMCallerImpl.java

```diff
@@ -283,6 +283,12 @@ public boolean onLeaderStop(final Status status) {
         });
     }
 
+    @Override
+    public boolean onBeforeLeaderStart() {
+        doBeforeLeaderStart();
```

Comment:
It seems to me that the fact, that we call `doBeforeLeaderStart()` directly, but not through the `enqueueTask` mechanism, which is done for all leader callbacks, could lead to to the linearisability problems. For example, when we transfer leadership, we call `org.apache.ignite.raft.jraft.core.NodeImpl#onLeaderStop` , but if this transferring timeouts, we return back the old leader, and we expect that `onBeforeLeaderStart` will be called, but because we call it directly, `onLeaderStop` could be invoked after `onBeforeLeaderStart`